src/TableFunctions/TableFunctionS3Distributed.cpp

```diff
@@ -0,0 +1,143 @@
+
+namespace DB
+{
+
+namespace ErrorCodes
+{
+    extern const int LOGICAL_ERROR;
+    extern const int NUMBER_OF_ARGUMENTS_DOESNT_MATCH;
+}
+
+void TableFunctionS3Distributed::parseArguments(const ASTPtr & ast_function, const Context & context)
+{
+    /// Parse args
+    ASTs & args_func = ast_function->children;
+
+    if (args_func.size() != 1)
+        throw Exception("Table function '" + getName() + "' must have arguments.", ErrorCodes::LOGICAL_ERROR);
+
+    ASTs & args = args_func.at(0)->children;
+
+    if (args.size() < 4 || args.size() > 7)
+        throw Exception("Table function '" + getName() + "' requires 4 to 7 arguments: cluster, url," + 
+            "[access_key_id, secret_access_key,] format, structure and [compression_method].",
+            ErrorCodes::NUMBER_OF_ARGUMENTS_DOESNT_MATCH);
+
+    for (auto & arg : args)
+        arg = evaluateConstantExpressionOrIdentifierAsLiteral(arg, context);
+
+    cluster_name = args[0]->as<ASTLiteral &>().value.safeGet<String>();
+    filename_or_initiator_hash = args[1]->as<ASTLiteral &>().value.safeGet<String>();
+
+    if (args.size() < 5)
+    {
+        format = args[2]->as<ASTLiteral &>().value.safeGet<String>();
+        structure = args[3]->as<ASTLiteral &>().value.safeGet<String>();
+    }
+    else
+    {
+        access_key_id = args[2]->as<ASTLiteral &>().value.safeGet<String>();
+        secret_access_key = args[3]->as<ASTLiteral &>().value.safeGet<String>();
+        format = args[4]->as<ASTLiteral &>().value.safeGet<String>();
+        structure = args[5]->as<ASTLiteral &>().value.safeGet<String>();
+    }
+
+    if (args.size() == 5 || args.size() == 7)
+        compression_method = args.back()->as<ASTLiteral &>().value.safeGet<String>();
+}
+
+
+ColumnsDescription TableFunctionS3Distributed::getActualTableStructure(const Context & context) const
+{
+    return parseColumnsListFromString(structure, context);
+}
+
+StoragePtr TableFunctionS3Distributed::executeImpl(
+    const ASTPtr & ast_function, const Context & context,
+    const std::string & table_name, ColumnsDescription /*cached_columns*/) const
+{
+    UInt64 max_connections = context.getSettingsRef().s3_max_connections;
+
+    /// Initiator specific logic
+    while (context.getClientInfo().query_kind == ClientInfo::QueryKind::INITIAL_QUERY)
+    {
+        auto poco_uri = Poco::URI{filename_or_initiator_hash};
+
+        /// This is needed, because secondary query on local replica has the same query-id
+        if (poco_uri.getHost().empty() || poco_uri.getPort() == 0)
+            break;
+
+        S3::URI s3_uri(poco_uri);
+        StorageS3::ClientAuthentificaiton client_auth{s3_uri, access_key_id, secret_access_key, max_connections, {}, {}};
+        StorageS3::updateClientAndAuthSettings(context, client_auth);
+
+        auto lists = StorageS3::listFilesWithRegexpMatching(*client_auth.client, client_auth.uri);
+        Strings tasks;
+        tasks.reserve(lists.size());
```

Comment:
It would be nice to add possibility to read s3 file in parallel, on different replicas.
If it is difficult for now, we should at least think how to add it later.
For now, protocol does not support it...

Reply:
It's always possible to create new cluster definition when all servers defined as shards, it shouldn't break anything because as i understand files being passed to servers in random fashion.

Reply:
True. And also skip_unavailable_shards becomes safe.

Reply:
It would be better to name this function `s3ClusterAllReplicas`. Because it takes cluster and use each replica of each shard as a standalone node. There is no replicas and shards.